tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -260,82 +371,77 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
-      addTree(classTreeCache.moduleAccessor.getOrElseUpdate(
+      addToMainBase(classTreeCache.moduleAccessor.getOrElseUpdate(
           classEmitter.genModuleAccessor(linkedClass)))
-  }
 
-  /** Emits the static fields of a linked class.
-   *
-   *  They are initialized with the zero of their type at this point. It is
-   *  the job of static initializers to properly initialize them.
-   */
-  private def emitLinkedClassStaticFields(linkedClass: LinkedClass,
-      builder: JSTreeBuilder): Unit = {
+    // Static fields
 
-    if (!linkedClass.kind.isJSType) {
+    val staticFields = if (linkedClass.kind.isJSType) {
+      Nil
+    } else {
       val classCache = getClassCache(linkedClass.ancestors)
       val classTreeCache = classCache.getCache(linkedClass.version)
 
-      builder.addJSTree(classTreeCache.staticFields.getOrElseUpdate(
-          classEmitter.genCreateStaticFieldsOfScalaClass(linkedClass)(classCache)))
+      classTreeCache.staticFields.getOrElseUpdate(
+          classEmitter.genCreateStaticFieldsOfScalaClass(linkedClass)(classCache))
     }
-  }
 
-  /** Emits the static initializer of a linked class, if any. */
-  private def emitLinkedClassStaticInitializer(linkedClass: LinkedClass,
-      builder: JSTreeBuilder): Unit = {
+    // Static initialization
 
-    if (!linkedClass.kind.isJSType)
-      builder.addJSTree(classEmitter.genStaticInitialization(linkedClass))
-  }
+    val staticInitialization = if (linkedClass.kind.isJSType) {
+      Nil
+    } else {
+      classEmitter.genStaticInitialization(linkedClass)
+    }
 
-  /** Emits the class exports of a linked class.
-   *
-   *  This is done after everything else has been emitted for all the classes
-   *  in the program. That is necessary because class exports can call class
-   *  value accessors, which may have unknown circular references.
-   */
-  private def emitLinkedClassClassExports(linkedClass: LinkedClass,
-      builder: JSTreeBuilder): Unit = {
+    // Class exports
 
-    /* `if` to avoid looking up the caches for nothing. Probably worth doing
-     * because only few classes have class exports.
-     */
-    if (linkedClass.classExports.nonEmpty) {
-      val classCache = getClassCache(linkedClass.ancestors)
-      val classTreeCache = classCache.getCache(linkedClass.version)
-
-      builder.addJSTree(classTreeCache.classExports.getOrElseUpdate(
-          classEmitter.genClassExports(linkedClass)(classCache)))
+    val classExports = if (linkedClass.classExports.isEmpty) {
+      Nil
+    } else {
+      val treeWithGlobals = classTreeCache.classExports.getOrElseUpdate(
+          classEmitter.genClassExports(linkedClass)(classCache))
+      addGlobalRefs(treeWithGlobals.globalVarNames)
+      treeWithGlobals.value
     }
+
+    // Build the result
+
+    new GeneratedClass(
+        main.reverse,
+        staticFields,
+        staticInitialization,
+        classExports,
+        mentionedDangerousGlobalRefs
+    )
```

Comment:
Because here nothing is cached. If we use `WithGlobals`, we pay the price of its cost every single time. That's why, in general, I avoid `WithGlobals` in `Emitter.scala`. It also brings much less advantages here. Worse, I expect the code to become *less* readable in this method with `WithGlobals` than without.

Overall, the ration benefit/cost is just not worth it here, like it is in `ClassEmitter`.